src/query/expression/src/schema.rs

```diff
@@ -75,6 +76,11 @@ pub struct DataField {
 pub struct TableSchema {
     pub(crate) fields: Vec<TableField>,
     pub(crate) metadata: BTreeMap<String, String>,
+
+    // define new fields as Option for compatibility
+    pub next_column_id: Option<u32>,
+    column_id_map: Option<BTreeMap<String, u32>>,
+    column_id_set: Option<HashSet<u32>>,
```

Comment:
what's `column_id_set` ?

Is this set can be generated from  `fields` and `column_id_map` ?

Reply:
yes, `column_id_set` can be generated from fields and column_id_map ?

Reply:
So I suggest to have a method to get the id set,  the `column_id_set` should not be stored into meta because it may cause inconsistent if set is different to map.

Reply:
Ok, tag `column_id_set` as `skip_serializing`, every time `TableSchema` init, it will be recalculated.

Reply:
Can we have a uniq id in `TableField` ? I think keeping a `Hashmap` in dataschema is too heavy.

Vec will outperform Hashmap in less than hundreds of elements.
And `TableField` can represent the real meaning of the mapping between `name` and `id`  

```
#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct TableField {
    name: String,
    default_expr: Option<String>,
	id: ColumnID,
    data_type: TableDataType,
}

```

Reply:
`BlockMeta.col_metas` has to save what column saved in this block, a field-unique id can not work, only schema-unique id can work.

Reply:
> a field-unique id can not work, only schema-unique id can work.

It is not field-unique, it is still schema scope unique. The same way to build field ids a s it does with a `column_id_map`.

Every piece of information you can extract from `column_id_map` can also be extracted from Fields, does it?

With a `column_id_map`, you'll get a Field instance from `fields: Vec<Field>`, by column-name, and get field id from `column_id_map`, by column-name, too.
Without a `column_id_map`, you can just get a Field instance with field-id directly from `fields: Vec<Field>`, right?

Reply:
Do we need to support modifying the columns inside the `Tuple`? If not, it is possible to just give the top-level `TableField` a `ColumnId` and the inner columns can be found via the `ColumnId` plus a path vector. This would simplify the complexity of the design.

Reply:
> Every piece of information you can extract from `column_id_map` can also be extracted from Fields, does it?
> 
> With a `column_id_map`, you'll get a Field instance from `fields: Vec<Field>`, by column-name, and get field id from `column_id_map`, by column-name, too. Without a `column_id_map`, you can just get a Field instance with field-id directly from `fields: Vec<Field>`, right?

I see. It is a recursive structure of TableDataType, not a recursive TableField, we can not add a field-id into TableDataType.

Then I do not have any question about this.

Reply:
If the inner column need to be modified, it's better to create another  `Tuple Type`, so all the nested column should have a new ColumnID.

E.g.   Add a Tuple field `Tuple { a, b , c }`, this will assign 3 ids  ----`n, n+1, n+2`, we just need to store the first ID  in 
`DataFiled {  Tuple,  ColumnID:   n }`

Reply:
> If the inner column need to be modified, it's better to create another Tuple Type, so all the nested column should have a new ColumnID.

Is it acceptable at storage level? Assumes that field `t Tuple {a,b}` has column id 1,2; If a user removes tuple field `t:a`, it's going to have to assign a new column id: `t Tuple{b}` with column-id 3, and update the column-id for every parquet file?

Reply:
> If the inner column need to be modified, it's better to create another Tuple Type, so all the nested column should have a new ColumnID.

Is it acceptable at storage level? Assumes that field `t Tuple {a,b}` has column id 1,2; If a user removes tuple field `t:a`, it's going to have to assign a new column id: `t Tuple{b}` with column-id 3, and update the column-id for every parquet file?

Reply:
I think only allow creating a brand new nested type is simpler.

That means, `t Tuple {a,b}` and `t Tuple{b}` are two different fields. Users can not query the old `t Tuple {a, b}` after changing it to `t Tuple{b}`.

Reply:
> Is it acceptable at storage level? Assumes that field `t Tuple {a,b}` has column id 1,2; If a user removes tuple field `t:a`, it's going to have to assign a new column id: `t Tuple{b}` with column-id 3, and update the column-id for every parquet file?

we are not going to update the removed column in parquet file. If Tuple `t` is removed, column id 1,2 are not used anymore, any parquet file with these two ids is not used to read this `t`.